snapshotter/volume_integ_test.go

```diff
@@ -0,0 +1,145 @@
+	"github.com/firecracker-microvm/firecracker-containerd/internal/integtest"
+	"github.com/firecracker-microvm/firecracker-containerd/proto"
+	"github.com/firecracker-microvm/firecracker-containerd/runtime/firecrackeroci"
+	"github.com/firecracker-microvm/firecracker-containerd/snapshotter/internal/integtest/stargz/fs/source"
+	"github.com/firecracker-microvm/firecracker-containerd/volume"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+const mib = 1024 * 1024
+
+func TestGuestVolumeFrom_Isolated(t *testing.T) {
+	integtest.Prepare(t, integtest.WithDefaultNetwork())
+	const (
+		vmID     = "default"
+		postgres = "docker.io/library/postgres:14.3"
+		alpine   = "docker.io/library/alpine:3.10.1"
+		runtime  = "aws.firecracker"
+	)
+
+	ctx := namespaces.WithNamespace(context.Background(), vmID)
+
+	client, err := containerd.New(integtest.ContainerdSockPath, containerd.WithDefaultRuntime(runtime))
+	require.NoError(t, err, "unable to create client to containerd service at %s, is containerd running?", integtest.ContainerdSockPath)
+	defer client.Close()
+	fcClient, err := integtest.NewFCControlClient(integtest.ContainerdSockPath)
+	require.NoError(t, err, "Failed to create fccontrol client")
+
+	vs := volume.NewSet(runtime)
+
+	// Add a non-stargz image with volumes.
+	localImage := volume.FromImage(client, postgres, "postgres-snapshot", volume.WithSnapshotter("devmapper"))
+	err = vs.AddFrom(ctx, localImage)
+	require.NoError(t, err)
+
+	// Add a stargz image.
+	// The volume directories must be specified since the host's containerd doesn't know about the image.
+	remoteImage := volume.FromGuestImage(
+		client, vmID, al2stargz, "al2-snapshot", []string{"/etc/yum"},
+		volume.WithSnapshotter("demux"),
+	)
+	mount, err := vs.PrepareDriveMount(ctx, 10*mib)
+	require.NoError(t, err)
+
+	_, err = fcClient.CreateVM(ctx, &proto.CreateVMRequest{
+		VMID: vmID,
+		RootDrive: &proto.FirecrackerRootDrive{
+			HostPath: "/var/lib/firecracker-containerd/runtime/rootfs-stargz.img",
+		},
+		NetworkInterfaces: []*proto.FirecrackerNetworkInterface{
+			{
+				AllowMMDS: true,
+				CNIConfig: &proto.CNIConfiguration{
+					NetworkName:   "fcnet",
+					InterfaceName: "veth0",
+				},
+			},
+		},
+		MachineCfg: &proto.FirecrackerMachineConfiguration{
+			VcpuCount:  2,
+			MemSizeMib: 2048,
+		},
+		ContainerCount: 1,
+		DriveMounts:    []*proto.FirecrackerDriveMount{mount},
+	})
+	require.NoErrorf(t, err, "Failed to create microVM[%s]", vmID)
+	defer fcClient.StopVM(ctx, &proto.StopVMRequest{VMID: vmID})
+
+	_, err = fcClient.SetVMMetadata(ctx, &proto.SetVMMetadataRequest{
+		VMID:     vmID,
+		Metadata: fmt.Sprintf(dockerMetadataTemplate, "ghcr.io", noAuth, noAuth),
+	})
+	require.NoError(t, err, "Failed to configure VM metadata for registry resolution")
+
+	err = remoteImage.Pull(ctx,
+		containerd.WithImageHandlerWrapper(source.AppendDefaultLabelsHandlerWrapper(al2stargz, 10*mib)),
+	)
+	require.NoError(t, err)
```

Comment:
Since different snapshotters need different pull parameters, I have to expose this Pull method to take the pull parameters.

Reply:
Could pull options be a functional parameter to `FromGuestImage`?

Reply:
No. Seems Go methods only have one `opts ...Type`.

Reply:
Is this too hacky?
```
type imageProviderConfig struct {
	snapshotter string
        pullOpts []containerd.RemoteOpts
}

func WithPullOptions(opts ...containerd.RemoteOpts) ImageOpt {
    return func(c *imageProviderConfig) {
        c.pullOpts = opts
    }
}
```

Reply:
That's good!